### ISSUE TITLE
PAL-Linux-SGX returns 0 for bash -c 'exit -1'

### ISSUE THREAD

**Issue by vahldiek** [oscarlab/graphene#1364]
On Ubuntu 18.04 bash uses execve and replaces itself with the command specified behind -c (here exit -1). Since Linux-SGX cannot replace the execution of the current process, it creates a new process for exit -1 and starts. The current process of bash ends silently without any errors and does not wait for the child to finish. At this point the grant parent and the child have no relation. This was already highlighted by PR https://github.com/oscarlab/graphene/pull/1181.

**Comment by dimakuv** [comment]
The correct way is to check for a corner case during `execve()`:

1. If I have the parent (I was forked from another Graphene process), then continue with the currently available logic of rewiring my parent and my exec'd child to each other. After that I can safely terminate.

2. New logic: If I don't have the parent (I was started by the user from non-Graphene process, e.g., command line), then rewire all pipes to my exec'd child but keep myself alive and do `waitpid()` on the child.

**Comment by vahldiek** [comment]
Thanks @dimakuv. 

After further tests, I can confirm that the same behavior happens also on Ubuntu 16.04. @dimakuv we may want to check your hunch that bash on 16.04 has a different behavior than 18.04 to make sure that this is the problem. Starting next week I may have a couple of cycles to look into the untrusted PAL implementation.

**Comment by mkow** [comment]
I'm not sure if merging exit codes of Graphene and the user app is a good idea in the first place. I'd say that running `./pal_loader app` should return Graphene's exit code (i.e. whether anything inside Graphene's runtime failed, like e.g. missing app manifest) and if the user also wants to know the app's exit code, they should use a dedicated argument passed to `pal_loader` (e.g. something like `--write-exit-code-to-this-pipe=smth`). But that's probably a controversial opinion, so better wait for others to jump in and comment.

**Comment by vahldiek** [comment]
To clarify and avoid the exit code controversy, it is not so much about the exit code and rather about the behavior of emulating exec system calls that replace the current process. Since this is impossible in Linux SGX PAL, the current solution is to spawn a new process (the only alternative). Compared to the actual exec, we now have two processes the parent (exec'd) and the child (with the new program). 

Graphene's current implementation doesn't need the parent's process and hence, lets this process exit (with successful exit code). As a result though the grant parent thinks the program has finished its execution. 

The exit example does not fully show this behavior as it focuses on the exit code. Lets consider a different example: `bash -c 'ls'`.  The actual output is as follows:

```
$ SGX=1 ./pal_loader bash.manifest.sgx -c 'ls' 
....
// Graphene output
...
$ 
LS_OUTPUT
```

What should be a surprise here is that the bash input `$` starts before the LS_OUTPUT (showing files and directories as usual). This is a direct result of emulating exec by creating a new child process and exiting the parent process without waiting on the child. At this point the grant parent does not wait for the child to finish and continuous with its regular behavior (here ask for the next command). But wait, the child is still running and producing output (here LS_OUTPUT). 

With @dimakuv we suggest to change this behavior in such a way that the parent waits for the child before calling exit.

Please take into account that this is a special case which we think is only caused during execs that replace the current process. Other fork/exec system calls should behave correctly.

**Comment by mkow** [comment]
Ok, then +1 to fixing this (modulo exit code, as I wrote in my previous comment). I don't like the 2. solution proposed by @dimakuv, this sounds like something which can easily explode into tons of processes just waiting for their children in cases like bash (or maybe not? do you think we can face this problem in real-world apps?).

Maybe we could just remove this logic and execve in the same process? AFAIR @chiache long time ago said that this was implemented only to start from a clean state (so, less error-prone), not because it can't be implemented otherwise.

**Comment by dimakuv** [comment]
> AFAIR @chiache long time ago said that this was implemented only to start from a clean state (so, less error-prone), not because it can't be implemented otherwise.

I don't think this is correct. When a process does `execve()`, it has opened file descriptors which should stay opened (unless some are marked as CLOEXEC). Graphene virtualizes these files descriptors into "PAL handles", so Graphene needs some way to make these Pal handles (with underlying host FDs) survive `execve()`.

Currently, this is done via the checkpointing process: the parent forks a new child and the parent sends PAL handles to the child via `DkSendHandle/DkReceiveHandle`. There is no way to do this in the same process (note that storing the checkpoint to disk and restoring the child from it is not implemented).

**Comment by yamahata** [comment]
Is the problem, pal_loader should wait for a process created by exec emulation?
Or the first process of pal-sgx should wait when exec is emulated?

If we assume user always invoke graphene process via pal_loader, the issue is former.

pal_loader can wait for a pipe to be written as suggested by @mkow by read builtin command (or file or whatever we can wait for instead of wait(2) system call) so that exit code is written instead of tricking parent-child relationship of graphene process.

**Comment by dimakuv** [comment]
@yamahata We can assume that the user always invokes Graphene via `pal_loader`, if this simplifies this issue. I actually like this idea (I missed this when @mkow proposed it). I guess during `DkProcessExit()`, we may send some magic string with exit code to stdin and this will signal to `pal_loader` that the child (even if it became detached) is done. Is that what you are proposing?

**Comment by mkow** [comment]
IMO sounds good, but we need to remember to handle EOF nicely (this will happen if Graphene crashes before writing the exit code).

**Comment by yamahata** [comment]
yes. pal_loader creates file descriptor for anonymous pipe and pass it to pal-sgx.

**Comment by dimakuv** [comment]
@yamahata @mkow @vahldiek I cooked a PR to fix this issue: https://github.com/oscarlab/graphene/pull/1370.

It turned out to be easier than I expected. I just ask the "intermediate" process to stick around by waiting on the execve'd child (via `wait4()`). This way the host shell doesn't become detached, and this "intermediate" process doesn't use any resources (just sleeps until the child terminates).

This solution means that even without `pal_loader` Graphene will work as expected.

It's a bit hacky still, but this whole story with `execve()` emulation is hacky.

**Comment by dimakuv** [comment]
> I don't like the 2. solution proposed by @dimakuv, this sounds like something which can easily explode into tons of processes just waiting for their children in cases like bash (or maybe not? do you think we can face this problem in real-world apps?).

What I implemented in #1370 is solution 2 which Michal disliked :) I forgot to reply to that.

I think no sane application will do "exec bombs" by spawning tons of processes at the same time. And exec'ing processes one after the other is fine, since their "intermediate" processes will terminate together with the exec'd processes. Also, these "intermediate" processes literally do nothing, just lie around.

So I don't think this has any serious side effects in sane real-world apps.
